### DISCUSSION TITLE
Prompt-User command

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#4003] [Development]
NeoMutt users are an inventive bunch, coming up with complex macros to solve their daily problems.
One feature they often ask for is a user prompt to set a variable.

I suggest a new command:

```
prompt_user { config_var | my_var } [ "message" ]
```

e.g.

```
# Ask the user for a $from value
prompt_user from "Enter from address"
```

The entered value would be put into the variable directly (no escaping or quoting necessary).
Hitting <kbd>Ctrl-G</kbd> (`$abort_key`) would leave the variable unchanged.

We might also want some options to fine-tune the behaviour.
e.g.

- `-hidden` display `*`s to conceal the input
- `-onechar` only read one char
- `-anykey` wait for any key
- `-menu` interpret (k)ey letters
- `-file` perform file/dir auto-completion 

---

~If the idea is liked, I'll turned it into a 'good first issue'~

What do you think of the idea?

**Vote now**: :+1: :-1: (and add comments / suggestions)

**Comment by motorto** [comment]
Hey @flatcap  I am interested in taking this "problem" .

**Comment by flatcap** [comment]
The challenge can be broken down into:

- Create a Command
- Parse function
- Prompt the user
- Set the variable
- Return Errors

## Create a Command

There are two parts to this.
In [`mutt_commands.c`](https://github.com/neomutt/neomutt/blob/master/mutt_commands.c) add:

- an entry to the [`mutt_commands[]` array](https://github.com/neomutt/neomutt/blob/master/mutt_commands.c#L49),
- a parsing function, `parse_prompt()`

[`struct Command`](https://github.com/neomutt/neomutt/blob/master/mutt_commands.h#L45-L60) defines both the format of the array and the prototype of the parsing function.

Put the function in [`command_parse.c`](https://github.com/neomutt/neomutt/blob/master/command_parse.c) and a prototype in [`command_parse.h`](https://github.com/neomutt/neomutt/blob/master/command_parse.h)

## Parse function

When the user types in the command, our parse function will be called.
It's the function's job to parse the string and peform any actions required.

We'll start simply, with no options.  The string will have the format:

```
prompt_user variable "message"
```

Use [`mutt_extract_token()`](https://github.com/neomutt/neomutt/blob/master/init.c#L386-L394) to read one "token" at a time, e.g. the command, a prompt string.
The token will be read into a [`struct Buffer`](https://github.com/neomutt/neomutt/blob/master/mutt/buffer.h#L30-L38) (an expanding string type).
Variable `buf` may be used as temporary space.

We'll need to validate the number of parameters given and the name of the variable.
The user may set any NeoMutt config variable, or create any variable that begins `my_`

A simple way to validate a NeoMutt config variable is to use [`cs_subset_lookup()`](https://github.com/neomutt/neomutt/blob/master/config/subset.c#L162-L168).
It returns a handle which we'll use later.

## Prompt the user

The actual prompting of the user is quite simple.
Call [`mutt_get_field()`](https://github.com/neomutt/neomutt/blob/master/gui/curs_lib.h#L91) or [`mutt_buffer_get_field()`](https://github.com/neomutt/neomutt/blob/master/gui/curs_lib.h#L85)

If the user cancels the prompt, with <kbd>Ctrl-G</kbd>, then we stop.

## Set the variable

If the user gave a `my_` variable, set its value with [`myvar_set()`](https://github.com/neomutt/neomutt/blob/master/myvar.c#L86-L91).
Otherwise, use [`cs_subset_he_string_set()`](https://github.com/neomutt/neomutt/blob/master/config/subset.c#L365-L373).
This function could fail if the value isn't valid for the type.

## Return Errors

Finally, return a [`enum CommandResult`](https://github.com/neomutt/neomutt/blob/master/mutt_commands.h#L31-L40) to show the success.
Any messages in `err` will be displayed to the user

**Comment by flatcap** [comment]
## Getting Started

Here are a few links with useful info:

- [Newbie Tutorial](https://neomutt.org/dev/newbie-tutorial) 
- [How to Build NeoMutt](https://neomutt.org/dev/build)
- [Coding Style](https://neomutt.org/dev/code)
- [Easy Tasks](https://neomutt.org/dev/easy-tasks)
- [Various Useful Build Tools](https://neomutt.org/dev/analysis)

**Comment by danielg0** [comment]
Is there still interest in introducing this feature? I'm new to NeoMutt development, but have started to work on a fork implementing it (https://github.com/danielg0/neomutt)

**Comment by flatcap** [comment]
Hi @danielg0 

> Is there still interest in introducing this feature?

Yes, please!

> have started to work on a fork implementing it

Fantastic.
The code looks reasonable.  Does it work? :-)

If so, please can you create a Pull Request, then I'll give it a good test and review.
(and we can think about adding some docs)

If not, or if you still have questions, we can discuss it here,
or you can also find us on IRC: `#neomutt` on `irc.libera.chat` ([web client](https://kiwiirc.com/nextclient/irc.libera.chat/#neomutt))

Thanks!

**Comment by danielg0** [comment]
Hi @flatcap

> The code looks reasonable.  Does it work? :-)

Basic functionality is done - calling prompt_user with a variable and
(optionally) a prompt, getting input from the user and setting the
variable all work as well as some error checking.

Now working on adding those flags you mentioned initially before making
a PR. `-hidden`, `-onechar` and `-file` are fairly straightforward, but
was unsure of the best way to implement `-anykey` (thinking I might add
an extra anykey argument to `mutt_buffer_get_field()` so it terminates
after the first call to `self_insert()`?). I was also hoping you could
clarify what you meant by "interpret (k)ey letters"?

Thanks

**Comment by flatcap** [comment]
> Basic functionality is done

great

> best way to implement `-anykey`

Here in `mutt_buffer_enter_fname()` might be useful:
https://github.com/neomutt/neomutt/blob/fa6afd8cf6cbb786709d78d096a317312cdfeb3d/gui/curs_lib.c#L465-L470

> thinking I might add an extra anykey argument to `mutt_buffer_get_field()`

That's a possibility, but I'm currently trying (and failing) to refactor that function.

> clarify what you meant by "interpret (k)ey letters"

Ah!  That would be a nice advanced feature...
Allowing the user to create menus, e.g.
```
Which profile: (W)ork, (P)ersonal?
```
which would only allow 'w', 'p' or `<abort>` in response.

It's probably quite a bit of work, so let's get the basics done first
(but keep thinking about this :-)
